Add library doc to guides (#27581)
angular/angular: aio/content/guide/use-libs.md

# Using Published Libraries

> Angular supplies some sophisticated libraries, such as [Angular Material](https://material.angular.io/). See the [Angular Resources](https://angular.io/resources) page for links to these, as well as third-party libraries, such as:

Reviewer: ```suggestion
When building Angular applications you can take advantage of sophisticated first-party libraries, such as [Angular Material](https://material.angular.io/), as well as rich ecosystem of third-party libraries. See the [Angular Resources](https://angular.io/resources) page for links to the most popular ones.
```

* [AngularFire](https://github.com/angular/angularfire2/blob/master/docs/install-and-setup.md)
* [FontAwesome](https://fontawesome.com/how-to-use/on-the-web/using-with/angular)

> * [Bootstrap 4](https://ng-bootstrap.github.io/#/home)

Reviewer: please delete the list - it duplicates the stuff on the Resources page.

> Published libraries typically have their own documentation, including installation instructions.

Reviewer: this is a general knowledge and not a valuable piece of info. delete?

## Installing libraries

> To use a published Angular library, use the [npm package manager](guide/npm-packages) or [yarn](https://yarnpkg.com/) to install it, then import it into your project code.

Reviewer: please rewrite this whole section to say that people should use `ng add` to install libraries. `ng add` will use `npm` or `yarn` under the hood depending on the project configuration. (unlike plane `npm install` or `yarn add`, `ng add` also invokes schematics to generate library specific scaffolding into the project. This is why `ng add` is preferred over `npm install` or `yarn add` - it might be good to create a callout section with this info)

<code-example format="." language="bash">
npm install <lib_name> --save
</code-example>

&mdash;or&mdash;

<code-example format="." language="bash">
yarn add <lib_name>
</code-example>

To integrate reusable library code into an application, you need to import the provided functionality where you will use it.
A published library should provide a README or other documentation on how to add that lib to your app.
For an example, see [Angular Material](https://material.angular.io/) docs.

> * For a library that contains Angular installation schematics, you can use the CLI command `ng add <lib_name>` to automatically perform all needed setup in your app, such as adding import statements, fonts, themes, and so on.

Reviewer: developers don't know upfront if a package contains schematics or not, this is why it's safer to always use `ng add`. (see my comment above)

 For example, an `ng add` schematic was added to the service worker library in v7. For an example of adding it to an app using the CLI, see the [v7 service worker docs](https://angular.io/guide/service-worker-getting-started#adding-a-service-worker-to-your-project).

> * If a library does not contain an `ng add` schematic, you need to add the import statements and other code by hand. For an example, see the [v5 service worker docs](https://v5.angular.io/guide/service-worker-getting-started#adding-a-service-worker-to-a-new-application).

Reviewer: this is not very valuable info, developers know how to use npm packages and import code from them. I would just delete this paragraph.

### Library typings

Library packages often include typings in `.d.ts` files; see examples in `node_modules/@angular/material`. If your library's package does not include typings and your IDE complains, you may need to install the library's associated `@types/<lib_name>` package.

> To configure a library that does not include typings in the same package, install the related `@types` package with npm, then open the TypeScript configuration file, `src/tsconfig.app.json`, and add the library to the `types` array.

Reviewer: adding these into "types" array is not strictly necessary, the cli project is configured to automatically lookup the `@types` packages if they are installed. please remove the part about modifying the configuration file.

> TypeScript looks for types in the `node_modules/@types` folder by default, so you don't have to add each type package individually.

Reviewer: yup.. just what I said above..

For example, suppose you have a library named `d3`:

<code-example format="." language="bash">
npm install d3 --save
npm install @types/d3 --save-dev
</code-example>

Types defined in the library need to be added to the TypeScript configuration for the project that uses it.

* Add the library to the "types" array in `src/tsconfig.app.json`.

```
"types":[
  "d3"
]
```

> * If the library is used only for e2e tests, add the library to the types array in  `e2e/tsconfig.e2e.json`.

Reviewer: delete

> * If the library is used for unit tests add to the types array in `src/tsconfig.spec.json`.

Reviewer: delete

> If a library doesn't have typings available at `@types/`, you can still use it by manually adding typings for it.

Reviewer: everything from here until the next section "maintaining libraries" is a bit wonky - it's incomplete and not that useful. could we just link to a third part docs on this? this one looks quite reasonable: https://medium.com/@chris_72272/migrating-to-typescript-write-a-declaration-file-for-a-third-party-npm-module-b1f75808ed2

Reviewer: I just saw what you did later in the section on how to add custom typings for packages loaded as global at runtime. If you provide similar examples here (e.g. declare an interface and then show how to use it) then this section should stay.

Author: @blove gave me an example -- @IgorMinar Please confirm if this is what you meant.

To do this:

1. Create a `typings.d.ts` file in your `src/` folder. This file is automatically included as global type definition.

1. Add the following code in `src/typings.d.ts`.
```
declare module 'typeless-package';
```

1. In the component or file that uses the library, add the following code:
```
import * as typelessPackage from 'typeless-package';
typelessPackage.method();
```

You can define more typings as needed.

> ## Maintaining libraries

Reviewer: ```suggestion
## Updating libraries
```

Libraries can be updated by their publishers, and also have their own dependencies which need to be kept current.
To check for updates to your installed libraries, use the `ng update` command.

* Use `ng update <lib_name>` to update individual library versions. The Angular CLI checks the latest published release of the library, and if the latest version is newer than your installed version, downloads it and updates your `package.json` to match the latest version.

> * Use the `--all` option to check all installed dependencies for the latest published versions.

Reviewer: I don't like to recommend --all especially to novice users because it can make too many changes at once and get the app into an unpredictable state if one of the updates contains an undocumented breaking change or a regression. I'd leave this --add sentence out. It's more suitable for the `ng update` doc.

Author: Leaving it out, but adding a link to the `ng update` doc page.

> When you update Angular to a new version, you need to make sure that any libraries you are using are current. If libraries have interdependencies, you might have to update them in a particular order.

Reviewer: what is this order? this paragraph is confusing. we should either delete it or point people to update.angular.io which has all the steps in the right order...

## Adding a library to the runtime global scope

> JavaScript libraries that are not imported into an app can be added to the runtime global scope and loaded as if they were in a script tag.

Reviewer: ```suggestion
Legacy JavaScript libraries that are not imported into an app can be added to the runtime global scope and loaded as if they were in a script tag.
```

npm install bootstrap --save
</code-example>

In the `angular.json` configuration file, add the associated script files to the "scripts" array:

```
"scripts": [
  "node_modules/jquery/dist/jquery.slim.js",
  "node_modules/popper.js/dist/umd/popper.js",
  "node_modules/bootstrap/dist/js/bootstrap.js"
],
```

Add the Bootstrap CSS file to the "styles" array:

```
"styles": [
  "node_modules/bootstrap/dist/css/bootstrap.css",
  "src/styles.css"
],
```

Run or restart `ng serve` to see Bootstrap 4 working in your app.

### Using runtime-global libraries inside your app

Once you import a library using the "scripts" array, you should **not** import it using an import statement in your TypeScript code (such as `import * as $ from 'jquery';`).
If you do, you'll end up with two different copies of the library: one imported as a global library, and one imported as a module.
This is especially bad for libraries with plugins, like JQuery, because each copy will have different plugins.

Instead, download typings for your library (`npm install @types/jquery`) and follow the library installation steps. This gives you access to the global variables exposed by that library.

> ### Defining typings for runtime-global libraries

Reviewer: aha.. this is good. now I understand why you had the previous section for custom typings above. I'll update my comment up.

A published library should provide a README or other documentation on how to add that lib to your app.
For an example, see [Angular Material](https://material.angular.io/) docs.

* For a library that contains Angular installation schematics, you can use the CLI command `ng add <lib_name>` to automatically perform all needed setup in your app, such as adding import statements, fonts, themes, and so on.

 For example, an `ng add` schematic was added to the service worker library in v7. For an example of adding it to an app using the CLI, see the [v7 service worker docs](https://angular.io/guide/service-worker-getting-started#adding-a-service-worker-to-your-project).

* If a library does not contain an `ng add` schematic, you need to add the import statements and other code by hand. For an example, see the [v5 service worker docs](https://v5.angular.io/guide/service-worker-getting-started#adding-a-service-worker-to-a-new-application).

### Library typings

To install a library that does not include typings, install the related `@types` with npm, then open `src/tsconfig.app.json` and add the library to the `types` array.

> Angular looks for types in the `node_modules/@types` folder by default, so you don't have to add each type individually.

Reviewer 2: It's actually TypeScript's decision.

<code-example format="." language="bash">
npm install jquery --save
npm install popper.js --save
npm install bootstrap --save
</code-example>

In the `angular.json` configuration file, add the associated script files to the "scripts" array:

```
"scripts": [
  "node_modules/jquery/dist/jquery.slim.js",
  "node_modules/popper.js/dist/umd/popper.js",
  "node_modules/bootstrap/dist/js/bootstrap.js"
],
```

Add the Bootstrap CSS file to the "styles" array:

```
"styles": [
  "node_modules/bootstrap/dist/css/bootstrap.css",
  "src/styles.css"
],
```

Run or restart `ng serve` to see Bootstrap 4 working in your app.

### Using runtime-global libraries inside your app

Once you import a library using the "scripts" array, you should **not** import it using an import statement in your TypeScript code (such as `import * as $ from 'jquery';`).
If you do, you'll end up with two different copies of the library: one imported as a global library, and one imported as a module.
This is especially bad for libraries with plugins, like JQuery, because each copy will have different plugins.

> Instead, download typings for your library (`npm install @types/jquery`) and follow the library Installation steps. This gives you access to the global variables exposed by that library.

Reviewer 2: ```suggestion
Instead, download typings for your library (`npm install @types/jquery`) and follow the library installation steps. This gives you access to the global variables exposed by that library.
```
